pkg/controller/podautoscaler/horizontal.go

```diff
@@ -502,6 +463,42 @@ func (a *HorizontalController) reconcileAutoscaler(hpav1Shared *autoscalingv1.Ho
 	return a.updateStatusIfNeeded(hpaStatusOriginal, hpa)
 }
 
+// normalizeDesiredReplicas takes the metrics desired replicas value and
+// normalizes it based on the appropriate conditions (i.e. < maxReplicas, >
+// minReplicas, etc...)
+func (a *HorizontalController) normalizeDesiredReplicas(hpa *autoscalingv2.HorizontalPodAutoscaler, currentReplicas int32, desiredReplicas int32) int32 {
+	// Do not upscale too much to prevent incorrect rapid increase of the number of master replicas caused by
+	// bogus CPU usage report from heapster/kubelet (like in issue #32304).
+	scaleUpLimit := calculateScaleUpLimit(currentReplicas)
```

Comment:
As I understand it, most of the problems we've faced with normalizing `desiredReplicas` relate to the code base applying only one of the normalization rules, when multiple were necessary. For example, if `desiredReplicas > scaleUpLimit > maxReplicas`, we previously set `desiredReplicas = scaleUpLimit`, and then did not check that `desiredReplicas <= maxReplicas`.

My opinion is that the simplest way to deal with this logic is to implement a recursive solution (see the code below), where we continuously apply normalization rules until we have a `desiredReplicas` value that passes all the rules.

However, I'm confused what to do about the `setCondition` messages. In the simplest example, imagine `desiredReplicas > maxReplicas`. In the first call to `normalizeDesiredReplicas`, we'll set the condition to be "the desired replica count was more than the maximum replica count". We'll then call `normalizeDesiredReplicas` with `maxReplicas` as the new `desiredReplicas`. This will pass all the rules - should we then set the condition to be "the desired replica count is within the acceptable range"? Or should there only be one `setCondition` set no matter how many times we recursively call `normalizeDesiredReplicas`?

Reply:
I see your point about checking all conditions. I'm a bit scared of recursion with complex stop condition though. For example if scaleUpLimit < hpa.Spec.MinReplicas this code would get into an endless recursion. It probably can't happen if scaleUpLimitFactor > 1 as we would go into a different branch in reconcileAutoscaler, but it makes me nervous nonetheless.

As to setting condition - at the end of the day a condition has a single value. If you keep changing it the last one is the only one that matters.

I would consider an alternative approach to solving the complexity of this method. All the cases either increase or decrease desiredReplicas. Why not just calculate minimum and maximum allowed value (respectively "MinReplicas if defined" || 0 and min(scaleUpLimit, MaxReplicas)) and compare desiredReplicas against that? Also we can compare minimum and maximum for sanity check if required.

Reply:
Ah that is very interesting  @MaciekPytel. At first, I didn't think that it would ever be possible for `scaleUpLimit` to be a value such that we infinitely recurse (because its based on the previous `replicas` value, which should be within the min/max limits). However, as I consider it more, I realize that its possible that we could set the number of replicas with one set of min/max replicas, and then update the hpa to a new set of min/max replicas. In that case, it is possible that we'd get stick in an infinite loop. Thanks for your careful eyes on this - you make an excellent point :)

Calculating minimum/maximum allowed value seems like a great option. I'll explore that.